x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,520 @@
+            funds: 300,
+        };
+
+        let res = execute(deps.as_mut(), mock_env(), info.clone(), send_msg.clone()).unwrap();
+        let Attribute { key, value } = &res.attributes[2];
+        assert_eq!(key, "weekly_used");
+        assert_eq!(value, "300");
+
+        let res = execute(deps.as_mut(), mock_env(), info.clone(), recv_msg.clone()).unwrap();
+        let Attribute { key, value } = &res.attributes[2];
+        assert_eq!(key, "weekly_used");
+        assert_eq!(value, "0");
+
+        // We can still use the channel. Even if we have sent more than the
+        // allowance through the channel (900 > 3000*.1), the current "balance"
+        // of inflow vs outflow is still lower than the channel's capacity/quota
+        let res = execute(deps.as_mut(), mock_env(), info.clone(), recv_msg.clone()).unwrap();
+        let Attribute { key, value } = &res.attributes[2];
+        assert_eq!(key, "weekly_used");
+        assert_eq!(value, "300");
+
+        let err = execute(deps.as_mut(), mock_env(), info.clone(), recv_msg.clone()).unwrap_err();
+
+        assert!(matches!(err, ContractError::RateLimitExceded { .. }));
+        //assert_eq!(18, value.count);
+    }
+
+    #[test]
+    fn asymetric_quotas() {
+        let mut deps = mock_dependencies();
+
+        let quota = QuotaMsg::new("weekly", RESET_TIME_WEEKLY, 10, 1);
+        let msg = InstantiateMsg {
+            gov_module: Addr::unchecked(GOV_ADDR),
+            ibc_module: Addr::unchecked(IBC_ADDR),
+            channels: vec![Channel {
+                name: "channel".to_string(),
+                quotas: vec![quota],
+            }],
+        };
+        let info = mock_info(GOV_ADDR, &vec![]);
+        let _res = instantiate(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+
+        // Sending 2%
+        let msg = ExecuteMsg::SendPacket {
+            channel_id: "channel".to_string(),
+            channel_value: 3_000,
+            funds: 60,
+        };
+        let info = mock_info(IBC_ADDR, &vec![]);
+        let res = execute(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+        let Attribute { key, value } = &res.attributes[2];
+        assert_eq!(key, "weekly_used");
+        assert_eq!(value, "60");
+
+        // Sending 1% more. Allowed, as sending has a 10% allowance
+        let msg = ExecuteMsg::SendPacket {
+            channel_id: "channel".to_string(),
+            channel_value: 3_000,
+            funds: 30,
+        };
+
+        let info = mock_info(IBC_ADDR, &vec![]);
+        let res = execute(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+        let Attribute { key, value } = &res.attributes[2];
+        assert_eq!(key, "weekly_used");
+        assert_eq!(value, "90");
+
+        // Receiving 1% should fail. 3% already executed through the channel
+        let recv_msg = ExecuteMsg::RecvPacket {
+            channel_id: "channel".to_string(),
+            channel_value: 3_000,
+            funds: 30,
+        };
```

Comment:
This feels incorrect right? (Relative to what we want)

E.g. the 1% recv here should be going to `-1%` of what we had the prior week, rather than total net change

Reply:
The way it's currently implemented it's tracking only one *balance*.  In that case, 3% has moved out of the chain, so if 1% were to move in, we're be at a 2% *balance*, so that's not allowed. The opposite would work: with `quotas={send: 3%, recv: 2%)`, even if you've already sent 3%, you could receive 1% cause that would put the balance at 2%. 

We could change the implementation to what you're describing though. This probably makes more sense if the quotas aren't going to be symetric. I think the "balance" implementation is a leftover from when I had the same quota for send and receive.

Reply:
>  going to -1% of what we had the prior week

Right now we're not tracking "what we had the previous week". Instead, the percentages are calculated based on the value of the channel at the time of the request. The current implementation actually allows more than the specified percentage (as the value grows/shrinks with every send) but the actual max converges pretty quickly for small percentages. The advantage is that we avoid tracking the the network value for each channel.

Happy to change the implementation here. I think both make sense

Reply:
I've just pushed a change so that this behaves as you expected. It's a bit less clean, as we need to consider both in and out flows, but closer to what a user would expect.

Reply:
Added the change for "what we had the previous week" (caching the value in the quota on the first call or when it expires) in https://github.com/osmosis-labs/osmosis/pull/2408/commits/c2a72add8d1accaf188b7a61b296c885170dc3cc. 

I find it a bit confusing that the differences in the quotas could be quite large when you have multiple quotas for a path (denom+channel). Consider, for example, if there is the supply was 100 last week, but someone trades 1M and wants to ibc 1%. Even if the daily 1% limit would allow you, the weekly still would prevent you from sending this through IBC.

Maybe this is what we want. We can have a quick discussion about the tradeoffs and pick one

Reply:
oh thats a good point, that its super weird. Lets def discuss what makes most sense